src/software/thunderscope/robot_communication.py

```diff
@@ -72,29 +75,132 @@ def __init__(
         )
 
         self.send_estop_state_thread = threading.Thread(target=self.__send_estop_state)
-        self.run_world_thread = threading.Thread(target=self.run_world, daemon=True)
+        self.run_world_thread = threading.Thread(target=self.__run_world, daemon=True)
         self.run_primitive_set_thread = threading.Thread(
-            target=self.run_primitive_set, daemon=True
+            target=self.__run_primitive_set, daemon=True
         )
 
-        # only checks for estop if checking is not disabled
-        if not self.disable_estop:
+        # initialising the estop
+        # tries to access a plugged in estop. if not found, throws an exception
+        # if using keyboard estop, skips this step
+        self.estop_reader = None
+        self.estop_is_playing = False
+        # when the estop has just been stopped,
+        # we want to send a stop primitive once to all currently connected robots
+        self.should_send_stop = False
+
+        # only checks for estop if we are in physical estop mode
+        if self.estop_mode == EstopMode.PHYSICAL_ESTOP:
             try:
                 self.estop_reader = tbots_cpp.ThreadedEstopReader(
                     self.estop_path, self.estop_buadrate
                 )
             except Exception:
                 raise Exception("Could not find estop, make sure its plugged in")
```

Comment:
now that the helper checks the estop path exists, change the error message to `f"invalid estop found at location {estop_path}"`